server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -121,26 +120,50 @@ public void consumeResult(SearchPhaseResult result, Runnable next) {
     public SearchPhaseController.ReducedQueryPhase reduce() throws Exception {
         if (pendingMerges.hasPendingMerges()) {
             throw new AssertionError("partial reduce in-flight");
-        } else if (pendingMerges.hasFailure()) {
-            throw pendingMerges.getFailure();
+        }
+        Exception failure = pendingMerges.failure.get();
+        if (failure != null) {
+            throw failure;
         }
 
         // ensure consistent ordering
         pendingMerges.sortBuffer();
-        final TopDocsStats topDocsStats = pendingMerges.consumeTopDocsStats();
-        final List<TopDocs> topDocsList = pendingMerges.consumeTopDocs();
+        final TopDocsStats topDocsStats = pendingMerges.topDocsStats;
+        final int resultSize = pendingMerges.buffer.size() + (pendingMerges.mergeResult == null ? 0 : 1);
```

Comment:
Just for my understanding, this size is a best effort as the size might change as it is not in a synchronized block?

Reply:
Hmm actually I just computed it here because I figured it must be safe since `sortBuffer` is called above and that's safe. Looking at the code, it seems since we're past having any partial reduces in flight here, the synchronisation in this method isn't needed at all? Lets maybe leave that to the next step though :) I think I can do a clean follow-up that removes any notion of blocking from this logic based on this PR, no need to go too far in one go :) For now I'd leave it here because we had that barrier before on `getAggs` though to not take any chances.